### ISSUE TITLE
Use of SimpleDateFormat is not thread-safe

### ISSUE THREAD

**Issue by kgregory-rjmetrics** [cognitect/transit-java#15]
`AbstractParser` maintains a static `SimpleDateFormat` instance, which is used by `WriteHandlers`. `SimpleDateFormat` is [documented](http://docs.oracle.com/javase/8/docs/api/java/text/SimpleDateFormat.html) as being non-threadsafe, and in practice using Transit from multiple threads will cause the following exception:

```
Exception in thread "async-thread-macro-13" java.lang.RuntimeException: java.lang.ArrayIndexOutOfBoundsException: 98
	at com.cognitect.transit.impl.WriterFactory$1.write(WriterFactory.java:129)
	at cognitect.transit$write.invoke(transit.clj:149)
	...
Caused by: java.lang.ArrayIndexOutOfBoundsException: 98
	at sun.util.calendar.BaseCalendar.getCalendarDateFromFixedDate(BaseCalendar.java:454)
	at java.util.GregorianCalendar.computeFields(GregorianCalendar.java:2333)
	at java.util.GregorianCalendar.computeFields(GregorianCalendar.java:2248)
	at java.util.Calendar.setTimeInMillis(Calendar.java:1140)
	at java.util.Calendar.setTime(Calendar.java:1106)
	at java.text.SimpleDateFormat.format(SimpleDateFormat.java:955)
	at java.text.SimpleDateFormat.format(SimpleDateFormat.java:948)
	at java.text.DateFormat.format(DateFormat.java:336)
	at com.cognitect.transit.impl.WriteHandlers$TimeWriteHandler$1.rep(WriteHandlers.java:345)
```

There are a few solutions to this:

1) Use  a `ThreadLocal` to wrap the date formatter, as shown in line 339 [here](http://sourceforge.net/p/practicalxml/code/HEAD/tree/trunk/src/main/java/net/sf/practicalxml/XmlUtil.java).
2) Re-create the date formatter for each call. In my experience, date formatter instances are quite cheap to construct.
3) Assuming that write handlers are not shared between threads, create the date formatter instance once, in `TimeWriteHandler.getVerboseHandler()`.
4) Clearly document that Transit is not thread-safe.

Option 3 is imo the best, but I haven't studied the codebase sufficiently to proceed with it. So I'll create a patch that follows option 1 (we're currently blocked by this issue).

**Comment by puredanger** [comment]
This is definitely a bug that needs to be fixed.

**Comment by kgregory-rjmetrics** [comment]
At this point I'm scratching my head, because I have 100% repeatability when running from Clojure, but cannot cause a failure when running from Java. Have tried running on Linux machines with 2 and 4 cores, using thread counts from 4-32. JDK 1.7-079 in both cases.

The Clojure code that I'm using to replicate is here: https://gist.github.com/kgregory-rjmetrics/55e6722108b9174b9fe9

The Java test case is in my development branch here: https://github.com/kgregory-rjmetrics/transit-java/blob/dev-fix-dateformat/src/test/java/com/cognitect/transit/MultiThreadTest.java

**Comment by dchelimsky** [comment]
Thanks for the repro code. I've run the Clojure example several times without ever seeing the exception, but just because I can't see it doesn't mean it's not a bug. I'll be looking into this today and will follow up later.

**Comment by dchelimsky** [comment]
@kgregory-rjmetrics we can't guarantee that users won't share handlers across threads, so I think the ThreadLocal is the right approach. Thanks for all the suggested solutions.

**Comment by dchelimsky** [comment]
Just released transit-java-0.8.307 and transit-clj-0.8.283 with this fix. Takes a bit for them to make it to maven's search. I'll email the transit-format mailing list once I see them there.
